osgar/drivers/tankmaria.py

```diff
@@ -0,0 +1,149 @@
+                                round(math.degrees(heading)*100)])
+
+
+    def compute_pose(self, left, right):
+        """Update internal pose with 'dt' step"""
+        if left is None or right is None:
+            return False, None, None
+        x, y, heading = self.pose
+
+        metricL = ENC_SCALE * left  # dt is already part of ENC_SCALE
+        metricR = ENC_SCALE * right
+
+        dist = (metricL + metricR)/2.0
+        angle = (metricR - metricL)/WHEEL_DISTANCE
+
+        # advance robot by given distance and angle
+        if abs(angle) < 0.0000001:  # EPS
+            # Straight movement - a special case
+            x += dist * math.cos(heading)
+            y += dist * math.sin(heading)
+            #Not needed: heading += angle
+        else:
+            # Arc
+            r = dist / angle
+            x += -r * math.sin(heading) + r * math.sin(heading + angle)
+            y += +r * math.cos(heading) - r * math.cos(heading + angle)
+            heading += angle # not normalized
+        return True, (x, y, heading), (dist, angle)
+
+    def get_packet(self, buf):
+        s = buf.split(b'\n')
+        if len(s) < 2:
+            return None, buf
+        return s[0].strip(), b'\n'.join(s[1:])
+
+    def slot_desired_speed(self, timestamp, data):
+        self.desired_speed, self.desired_angular_speed = data[0]/1000.0, math.radians(data[1]/100.0)
+
+    def slot_raw(self, timestamp, data):
+        self.buf += data
+        packet, self.buf = self.get_packet(self.buf)
+        while packet is not None:
+            try:
+                d = [int(x) for x in packet.split()]
+            except ValueError:
+                print('Invalid packet', packet.split())
+                packet, self.buf = self.get_packet(self.buf)
+                continue
+
+            if self.verbose:
+                print(d)
+            assert len(d) == 4, d
+            prev = self.encoder
+            self.milisec = d[0]
+            self.encoder = d[1], d[2]  # more inputs expected
+            if prev is not None:
+                left, right = self.encoder[0] - prev[0], self.encoder[1] - prev[1]
+                self.publish('encoders', [left, right])
+                valid, pose, dist_angle = self.compute_pose(left, right)
+                if valid:
+                    self.pose = pose
+                self.send_pose()
+
+            packet, self.buf = self.get_packet(self.buf)
+
+    def slot_tick(self, timestamp, data):
+        STEP = int(min(255, abs(100 * self.desired_speed)))  # test PWM
+        if self.desired_speed > 0:
+            left, right = STEP, STEP
+        elif self.desired_speed < 0:
+            left, right = -STEP, -STEP
+        elif self.desired_angular_speed > 0:
+            left, right = -STEP, STEP
+        elif self.desired_angular_speed < 0:
+            left, right = STEP, -STEP
+        else:
+            left, right = 0, 0  # STOP
+
+        self.publish('raw', b'%d %d %d\n' % (left, right, self.led))
```

Comment:
So the protocol is text based and it expects that the first two integers are raw pwm values for left and right motor? And there is no provisioning for different command packets?

Reply:
ACK, the next version should have "structure based binary protocol" like was on Daisy and ATmega8board based robots.